descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h

```diff
@@ -0,0 +1,204 @@
+            uniformly sampled from either adjacent rung of the current state.
+            Note: there is no gaurentee that state sill be within the distance */
+        void sampleUniformNear(ompl::base::State *state, const ompl::base::State *near, double distance) override;
+
+        /** \brief Sample a state such that each component state[i] has
+            a Gaussian distribution with mean mean[i] and standard
+            deviation stdDev. If the sampled value exceeds the state
+            space boundary, it is thresholded to the nearest boundary. */
+        void sampleGaussian(ompl::base::State *state, const ompl::base::State *mean, double stdDev) override;
+    };
+
+    /** \brief A state space representing the Descartes boost graph. The distance function is determined by the difference
+     * in rungs and indices unless the rungs are adjacent, then an actual cost is made calculated using an edge evaluator */
+    template <typename FloatType>
+    class DescartesStateSpace : public ompl::base::StateSpace
+    {
+    public:
+
+        /** \brief The definition of a state in the Descartes State Space */
+        class StateType : public ompl::base::State
+        {
+        public:
+            /** \brief The definition of the vertex (rung, idx) */
+          std::pair<long unsigned int, long unsigned int> vertex;
+        };
+
+        /** \brief Constructor. The dimension of of the space needs to be specified. A space representing
+            the Descartes State Space will be instantiated */
+        DescartesStateSpace<FloatType>(descartes_light::BGLGraph<FloatType> graph,
+                                       std::vector<std::vector<VertexDesc<FloatType>>> ladder_rungs,
+                                       const std::vector<typename descartes_light::EdgeEvaluator<FloatType>::ConstPtr>& edge_eval,
+                                       const double max_dist,
+                                       const double rung_to_rung_dist = 100000)
+          : graph_(graph)
+          , ladder_rungs_(ladder_rungs)
+          , edge_eval_(std::move(edge_eval))
+          , max_dist_(max_dist)
+          , rung_to_rung_dist_(rung_to_rung_dist)
+          , stateBytes_(sizeof(std::pair<long unsigned int, long unsigned int>))
+        {
+            type_ = 14; // Larger than default types
+            setName("Descartes" + getName());
+        }
+
+        ~DescartesStateSpace() override = default;
+
+        /** \brief This state space is dicsrete which causes certain behaviors in OMPL */
+        bool isDiscrete() const override;
+
+        /** \brief Return the number of dimenstions in this space (2) */
+        unsigned int getDimension() const override;
+
+        /** \brief Get the maximum distance possible to record, this is the rung_to_rung_dist_*(number of rungs) */
+        double getMaximumExtent() const override;
+
+        /** \brief Returns the number of rungs in the ladder */
+        double getMeasure() const override;
```

Comment:
What exactly does this mean?

Reply:
Looking into this further I realize that I have implemented this wrong. It is supposed to represent the "volume" of the space. So I'll have to multiply the number of rungs by the length of a rung.